### ISSUE TITLE
Improve building with test coverage information

### ISSUE THREAD

**Issue by AlekSi** [FerretDB/FerretDB#2102]
### What should be done?

Go 1.20 improved Coverage profiling support for integration tests: https://go.dev/testing/coverage/ We should use it instead of our old hacks.

### Where?

Taskfiles, CI pipelines, etc.

### Definition of Done

- https://github.com/FerretDB/FerretDB/blob/main/cmd/ferretdb/main_test.go removed;
- `ferretdb_testcover` build tag is removed;
- coverage of integration tests is still collected as before.

**Comment by durgakiran** [comment]
hi @AlekSi  would like to work on this issue, can you please assign this to me.

**Comment by AlekSi** [comment]
Done, thank you!

**Comment by AlekSi** [comment]
@durgakiran Any news? Anything we could help with?

**Comment by durgakiran** [comment]
Facing some issues with setup, will try again in my new ubuntu.

**Comment by durgakiran** [comment]
@AlekSi Here are the steps I followed. Facing issues in test step.

1. task init
2. task env-pull
3. task env-up
4. task test

Here is the error message

```
ferretdb_postgres          | 2023-06-26 16:01:17.117 NDT [1013] ERROR:  schema "TestUpdateArrayCompatPull-Array_pg" does not exist
ferretdb_postgres          | 2023-06-26 16:01:17.117 NDT [1013] STATEMENT:  DROP SCHEMA "TestUpdateArrayCompatPull-Array_pg" CASCADE
ferretdb_postgres          | 2023-06-26 16:01:17.290 NDT [1017] ERROR:  schema "TestDeleteCompat-Two_pg" does not exist
ferretdb_postgres          | 2023-06-26 16:01:17.290 NDT [1017] STATEMENT:  DROP SCHEMA "TestDeleteCompat-Two_pg" CASCADE
ferretdb_postgres          | 2023-06-26 16:01:18.462 NDT [1021] ERROR:  schema "TestDeleteCompat-UnorderedError_pg" does not exist
ferretdb_postgres          | 2023-06-26 16:01:18.462 NDT [1021] STATEMENT:  DROP SCHEMA "TestDeleteCompat-UnorderedError_pg" CASCADE
ferretdb_postgres          | 2023-06-26 16:01:18.606 NDT [1025] ERROR:  schema "TestDeleteCompat-One_pg" does not exist
ferretdb_postgres          | 2023-06-26 16:01:18.606 NDT [1025] STATEMENT:  DROP SCHEMA "TestDeleteCompat-One_pg" CASCADE
```




Other errors I see in the test terminal

```
wire_test.go:136: 
        	Error Trace:	/Users/kiran/projects/FerretDB/internal/wire/wire_test.go:136
        	            				/Users/kiran/projects/FerretDB/internal/wire/op_query_test.go:109
        	            				/usr/local/go/src/testing/fuzz.go:722
        	            				/usr/local/go/src/runtime/asm_arm64.s:1172
        	Error:      	Received unexpected error:
        	            	[records_test.go:88 wire.loadRecords] ../../tmp/records/00/00e58bd132ce4dd24875fac15b7b56635bb596f7ed50cd4454433ed3c2f46c41.bin: [msg_body.go:78 wire.ReadMessage] [op_msg.go:225 wire.(*OpMsg).UnmarshalBinary] wire.OpMsg.Document: validation failed for { debugError: { NaN: nan.0 }, $db: "TestDebugError" } with: NaN is not supported
        	Test:       	FuzzQuery
--- FAIL: FuzzReply (5.07s)
    wire_test.go:136: 
        	Error Trace:	/Users/kiran/projects/FerretDB/internal/wire/wire_test.go:136
        	            				/Users/kiran/projects/FerretDB/internal/wire/op_reply_test.go:100
        	            				/usr/local/go/src/testing/fuzz.go:722
        	            				/usr/local/go/src/runtime/asm_arm64.s:1172
        	Error:      	Received unexpected error:
        	            	[records_test.go:88 wire.loadRecords] ../../tmp/records/00/00e58bd132ce4dd24875fac15b7b56635bb596f7ed50cd4454433ed3c2f46c41.bin: [msg_body.go:78 wire.ReadMessage] [op_msg.go:225 wire.(*OpMsg).UnmarshalBinary] wire.OpMsg.Document: validation failed for { debugError: { NaN: nan.0 }, $db: "TestDebugError" } with: NaN is not supported
        	Test:       	FuzzReply
FAIL
```

**Comment by durgakiran** [comment]
FYI, my system is MAC (m2 chip)

**Comment by rumyantseva** [comment]
Hey @durgakiran! So, the errors you face are generated by fuzzing. That would be wonderful if you could create a new issue, describe the error, and upload the file that caused it (`tmp/records/00/00e58bd132ce4dd24875fac15b7b56635bb596f7ed50cd4454433ed3c2f46c41.bin`), so that we can reproduce and fix it. 

Then to get unblocked, you can delete all the files from `tmp/records` and run tests again.

**Comment by durgakiran** [comment]
Thank you that unblocked me

**Comment by durgakiran** [comment]
I need some help here understanding how we are running integration tests, as per my understanding we run integration tests against a server (assuming it as ferretdb server). I don't see any build (go build) command being run before we run integration tests, then how are we spinning up ferretdb server? Also, what is this `ferretdb_testcover` actually doing?

**Comment by AlekSi** [comment]
When `-target-url` is empty, a test-specific FerretDB instance is started for that test. The code for that is there: https://github.com/FerretDB/FerretDB/tree/main/integration/setup. That enables test coverage collection by standard `go test` flags. But this issue is not about that.

When `-target-url` is not empty, the given MongoDB URI is used for the target system. There is no task for that, but that mode is supported.

`ferretdb_testcover` build tag wraps the main logic of FerretDB in a test: https://github.com/FerretDB/FerretDB/blob/main/cmd/ferretdb/main_test.go. That allows us to run FerretDB, connect to it with any client, run any tests we want (including integration tests with non-empty `-target-url`), and collect the test coverage.

**Comment by durgakiran** [comment]
I did `task env-up` which I can see setups all the  backend servers (postgres, mongodb etc) and then I added one more task
```
env-ferretdb-build:
    desc: "Run ferretdb proxy"
    cmds:
      - rm -rf covdatafiles
      - mkdir covdatafiles
      - go build -v                 -o=bin/ferretdbmanual -cover -race=true ./cmd/ferretdb
      - ./bin/ferretdbmanual
```
I added these two environment variables 
```
GOCOVERDIR: covdatafiles
 FERRETDB_POSTGRESQL_URL: postgres://username:password@127.0.0.1:5432/ferretdb
```
after running above task (`env-ferretdb-build:`) , I have ferretdb proxy running, Now if I run the task `test-integration-pg` It is not showing coverage percentage. I am neither seeing any logs coming from ferretdb proxy server nor I am not seeing codecoverage which is saying 0%.

I am not able to get how we are running integration tests are we running them against a single proxy server, or each test case creates a proxy ferretdb server and then executes them?


PS: I deleted main_test.go file from ferretdb.

**Comment by AlekSi** [comment]
@durgakiran please send a draft PR — it is much easier to discuss code where code is :)

**Comment by durgakiran** [comment]
Here is draft a pull request #3059 . I managed to hit the ferretdb proxy server when I used  `-target-url`. 

But some test cases are failing or getting timeout error. Here is the output [file](https://docs.google.com/document/d/1HWJej3GI47OsqTzZ3GmrZ1n6F0cY8Vq3-o0_jefQ7so/edit?usp=sharing) of the `test-integration-pg`.  Can someone help solve the errors in test cases. they used to work fine when I was not passing `-target-url`



Also I am not sure may be because of this error I am not getting the coverage of `cmd/ferretdb`.

```
github.com/FerretDB/FerretDB/build/version      coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/cmd/ferretdb       coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/backends  coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/backends/sqlite   coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/backends/sqlite/metadata       coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/backends/sqlite/metadata/pool  coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/bson      coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/clientconn        coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/clientconn/conninfo   coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/clientconn/connmetricscoverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/clientconn/cursor coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/handlers  coverage: [no statements]
        github.com/FerretDB/FerretDB/internal/handlers/common   coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/handlers/common/aggregations   coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/handlers/common/aggregations/operators coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/handlers/common/aggregations/operators/accumulators    coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/handlers/common/aggregations/stages    coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/handlers/common/aggregations/stages/projection coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/handlers/commoncommandscoverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/handlers/commonerrors coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/handlers/commonparams coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/handlers/pg       coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/handlers/pg/pgdb  coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/handlers/proxy    coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/handlers/registry coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/handlers/sjson    coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/handlers/sqlite   coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/types     coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/types/fjson       coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/util/contract     coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/util/ctxutil      coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/util/debug        coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/util/debugbuild   coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/util/iterator     coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/util/lazyerrors   coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/util/logging      coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/util/must coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/util/observability    coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/util/resource     coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/util/state        coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/util/telemetry    coverage: 0.0% of statements
        github.com/FerretDB/FerretDB/internal/wire      coverage: 0.0% of statements
```

**Comment by AlekSi** [comment]
Did you start the environment with `task env-up`? `http://127.0.0.1:14268/api/traces` post failures indicate that Jaeger is not running.

**Comment by durgakiran** [comment]
yes, but I just observed that one of my docker container is down apparently `jaeger`

**Comment by AlekSi** [comment]
`task env-down env-up` might help

**Comment by durgakiran** [comment]
But the issue I am taking about is about failed testcases. I am removing the `api/traces` calls from log file so that the test case error logs are more visible.

**Comment by durgakiran** [comment]
Here is a sample part of error message 
```
2023/07/17 22:44:45 Post "http://127.0.0.1:14268/api/traces": dial tcp 127.0.0.1:14268: connect: connection refused
panic: test timed out after 35m0s
running tests:
	TestAggregateCompatAddFields (32m36s)
	TestAggregateCompatGroup (32m37s)
	TestAggregateCompatOptions (32m35s)
	TestAggregateCompatProject (32m43s)
	TestAggregateCompatProject/DotNotationExcludeTwo (2s)
	TestAggregateCompatProject/DotNotationExcludeTwo/TestAggregateCompatProject_OverflowVergeDoubles (0s)
	TestAggregateCompatProject/Exclude1FieldInclude1Field (1s)
	TestAggregateCompatProject/Exclude1FieldInclude1Field/TestAggregateCompatProject_Scalars (0s)
	TestAggregateCompatProject/IDType (2s)
	TestAggregateCompatProject/IDType/TestAggregateCompatProject_Doubles (0s)
	TestAggregateCompatProject/IncludeFieldIncludeID (1s)

```

**Comment by AlekSi** [comment]
Something like `task test-integration-pg TEST_RUN=TestQuery` should be enough for this issue. No need to run all tests locally to check if test coverage is collected.

**Comment by durgakiran** [comment]
OK now I have coverage report being generated for postgres integration tests. I observed that mongodb we are sending actual mongdb url (running in container) as the target url, how can such scenario be actual tests for our proxy?

I am referring this task
```
test-integration-mongodb:
    desc: "Run integration tests for MongoDB"
    dir: integration
    cmds:
      - >
        go test -count=1 -run='{{or .TEST_RUN .SHARD_RUN}}' -timeout={{.TEST_TIMEOUT}} {{.RACE_FLAG}} -tags={{.BUILD_TAGS}} -shuffle=on -coverpkg=../...
        -coverprofile=integration-mongodb.txt .
        -target-url=mongodb://127.0.0.1:47017/
        -target-backend=mongodb
    vars:
      SHARD_RUN:
        sh: go run -C .. ./cmd/envtool tests shard --index={{.SHARD_INDEX}} --total={{.SHARD_TOTAL}}
```

**Comment by AlekSi** [comment]
This configuration tests MongoDB only. We gather coverage only to see what tests are run, if there is some dead code, etc. Other configuration (like `test-integration-pg`) either test FerretDB only (with PostgreSQL in that case) or (so-called compatibility tests) test both FerretDB and MongoDB.

There is no need to test FerretDB's proxy mode.

**Comment by durgakiran** [comment]
So I am in correct path by not updating `test-integration-mongodb`

**Comment by AlekSi** [comment]
Yep!
